### ISSUE TITLE
Handling fees of taking snapshot

### ISSUE THREAD

**Issue by jw-pyo** [Strawberry-Pizza/ICO2.0#3]
In current system, both tapVoting and refundVoting,
we take snapshots of every token holder's balance before finalizing voting.

When someone(maybe it's one of developers) call the snapshot function, it spends a huge amount of gas because we iterate all of the token holder's account. And the caller(dev) should bear entire fees. I think it is unfair and ineffective.
In order to avoid paying a huge fee by only one caller, each token holder should pay their snapshot fee. But it also has an issue: 

> if we make holders pay their own snapshot fee, each holder call the function on their own.

One solution is taking snapshots when **holders throw a vote** , but it cannot guarantee that the balance of snapshot time is equal to the balance of finishing voting time.

In short, there are two ways:

- function caller(Dev) bear all operation fee for taking snapshots
- taking snapshot each other when throwing a vote

And write down if you have any brilliant ideas!

**Comment by JhChoy** [comment]
2nd one is better.
1st one is a huge burden on developers. Also, incentive receivers can store their incentives.
For example, let's assume there are n receivers and 5 times tap withdrawals.
1st way : dev will pay n X 5 amount of gas price.
2nd way : process of withdrawing incentives will less than n X 5 because many of them will store the incentives because of gas price. In addition, whole gas fees are divided.

**Comment by jw-pyo** [comment]
In perspective of cost, it is right. But if we follow 2nd way, the time of taking snapshot will be different each other. In this case, we would change our model. We cannot trace individual token after taking snapshot.

**Comment by JhChoy** [comment]
we can store their incentives in contract.
`close` function called => `_snapshot` function called => finalize vote & store their incentives.
We can make storing process as same as locked tokens. (It would be mapping type)

Receivers can just withdraw incentives from locked incentive tokens.
If someone keeps his/her incentives, and after next tap withdraw, we just add amount of additional incentives to mapping.

**Comment by jw-pyo** [comment]
Um, I mean if we take snapshot in the moment of throwing a vote, it can occur tap attack. A large token holders vote(take snapshot) and transfer their token to other accounts and take snapshot, and so on.

In current coding, `_snapshot()` function is a global function, just called once by someone and caller burdens all of the gas cost. It is 1st way of which I suggested.

Receiving incentive function call is surely called by each receivers. I agree with that.

**Comment by JhChoy** [comment]
To solve snapshot problem :

We know the gas price of computing snapshot is high. Therefore, it is hard to compute `_snapshot()` and calculate q and v.
Then, we have to bypass the process.
1. When system needs computing `_snapshot()` or calculating q and v, developers do process on local. Because of the below validation process, the system pauses the voting.
2. Devs put the results in the system with blocknumber. (Events will occur)
3. Validators do same process on their local environments and compare with the results.
4. If same, there's nothing happens.
5. If not, validators can execute `challenge()` function which is 'multiparty-computation' function. For example, if we need 10000 process during snapshot, 1 challenge means about 100 process. So, if challenge function called 100 times, the whole computation would finish and then, compare with the results of devs.
7. When the result is different, part of Dev's ethers are slashed.
8. Validation participants get part of slashed ethers, and left of them are burned.
